### PR TITLE
Add Pile-Up Infrastructure at the Generator Level, Restructure Generation package

### PR DESCRIPTION
**Pull request by vvolkl**
Two new algorithms extending the current HepMC Reader and a (trivial) implementation of the IPileUpTool -- something that is closer to the actual planned collisions is on the agenda. The solution for the  container of the hepmc event pile is not final -- in terms of memory management it is fairly intransparent.

### CODE REVIEW

**Comment by vvolkl** [comment]
Will rewrite this PR with just one Algorithm `HepMCReader` where pileup is added and configured with an optional tool, and update the package to the agreed-upon style at the same time.

**Comment by jlingema** [comment]
retest this please

**Comment by jlingema** [comment]
Hi Valentin,

Thanks for cleaning up the Generation package! This is very nice...

I see that you have a couple of cases where you just added (unnecessary) whitespace. It would be great if you could try and avoid these. Apart from that I have a couple of small comments I'll do as inline comments in a second.

Cheers,
Joschka

File: Generation/options/particleGun.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+dumper = HepMCDumper("Dumper")
+dumper.DataInputs.hepmc.Path="hepmc"
+
+
+# A particle table file needs to be present in the directory 
+# from which Gaudi is run. If you run from the top-level directory,
+# the following option can be added to the service to use the file in this directory.
+#ParticlePropertiesFile='Generation/options/ParticleTable.txt'
+
+particlePropertySvc = Gaudi__ParticlePropertySvc("ParticlePropertySvc", ParticlePropertiesFile='/afs/cern.ch/lhcb/software/releases/GAUDI/GAUDI_v26r4/GaudiExamples/tests/data/ParticleTable.txt')
```
Comment (jlingema) [inline Generation/options/particleGun.py]: Why don't you use the local copy of the ParticleTable?

File: Generation/options/pileup_hepmcreader.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    xVertexMin = -0.5,
+    xVertexMax = 0.5,
+    yVertexMin = -0.5,
+    yVertexMax = 0.5,
+    zVertexMin = -35,
+    zVertexMax = 35 )
+
+
+genpileup = PoissonPileUp(name="Pileup", 
+    Filename="/afs/cern.ch/user/v/vavolkl/public/example_MyPythia.dat",
```
Comment (jlingema) [inline Generation/options/pileup_hepmcreader.py]: Can you point this to the fcc-sw dir:
`/afs/cern.ch/exp/fcc/sw/0.6/testsamples/example_MyPythia.dat`

File: Generation/options/pileup_hepmcreader.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    yVertexMax = 0.5,
+    zVertexMin = -35,
+    zVertexMax = 35 )
+
+
+genpileup = PoissonPileUp(name="Pileup", 
+    Filename="/afs/cern.ch/user/v/vavolkl/public/example_MyPythia.dat", 
+    numPileUpEvents=2)
+reader = HepMCReader("Reader", 
+    Filename="/afs/cern.ch/user/v/vavolkl/public/example_MyPythia.dat",
```
Comment (jlingema) [inline Generation/options/pileup_hepmcreader.py]: Same as above, point to fcc-sw dir.

File: Generation/src/components/ConstPileUp.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+#include "ConstPileUp.h"
+
+#include "GaudiKernel/DeclareFactoryEntries.h"
+
+DECLARE_TOOL_FACTORY( ConstPileUp )
+
+ConstPileUp::ConstPileUp( const std::string& type,
+                          const std::string& name,
+                          const IInterface* parent )
+    : GaudiTool ( type, name , parent ) {
```
Comment (jlingema) [inline Generation/src/components/ConstPileUp.cpp]: Please reformat according to coding style:
https://github.com/HEP-FCC/FCCSW/blob/master/doc/CppCodingStyleGuidelines.md#indentation

File: Generation/options/particleGun.py (hunk 1 above)
Comment (vvolkl) [inline Generation/options/particleGun.py]: Right, I wanted to ask you about that. CTest was failing even with WORKING_DIRECTORY set the path of the file. I'll check again.

File: Generation/src/components/HepMCFileReaderTool.cpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+#include "GaudiKernel/DeclareFactoryEntries.h"
+
+DECLARE_TOOL_FACTORY(HepMCFileReader)
+
+HepMCFileReader::HepMCFileReader(
+    const std::string& type,
+    const std::string& name,
+    const IInterface* parent)
+    : GaudiTool(type, name, parent), m_file(0) {
```
Comment (jlingema) [inline Generation/src/components/HepMCFileReaderTool.cpp]: Slight reformat to conform with coding conventions, see above.

File: Generation/src/components/HepMCFileReaderTool.cpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+HepMC::GenEvent* HepMCFileReader::readNextEvent() {
+  Assert(0 != m_file, "Invalid input file!");
+  HepMC::GenEvent* tmpEvent = new HepMC::GenEvent();
+  m_file->fill_next_event(tmpEvent);
+  return tmpEvent;
+}
+
+
+StatusCode HepMCFileReader::finalize() {
+  if ( 0 != m_file ) {
```
Comment (jlingema) [inline Generation/src/components/HepMCFileReaderTool.cpp]: Could you change this to nullptr? Same in previous uses above.

File: Generation/src/components/HepMCFileReaderTool.h [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+  /// Wrapper for HepMC's fill_next_event()
+  virtual HepMC::GenEvent* readNextEvent();
+  /// Wrapper for HepMC file io.
+  virtual StatusCode open(const std::string& filename);
+
+private:
+  void close();
+  std::string m_filename;
+  HepMC::IO_GenEvent* m_file;
```
Comment (jlingema) [inline Generation/src/components/HepMCFileReaderTool.h]: Can you change this to a smart pointer, i.e. `std::unique_ptr`?

File: Generation/src/components/HepMCFileReaderTool.h [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+class HepMCFileReader: public GaudiTool, virtual public IHepMCFileReaderTool {
+public:
+  HepMCFileReader( const std::string& type, const std::string& name,
+    const IInterface* parent) ;
+  virtual ~HepMCFileReader(); ///< Destructor
+  virtual StatusCode initialize();
+  virtual StatusCode finalize();
+
+  /// Wrapper for HepMC's fill_next_event()
+  virtual HepMC::GenEvent* readNextEvent();
```
Comment (jlingema) [inline Generation/src/components/HepMCFileReaderTool.h]: Can you add a comment about the lifetime of `HepMC::GenEvent`?

File: Generation/src/components/HepMCHistograms.cpp [hunk 9]
```diff
@@ -0,0 +1,68 @@
+#include "HepMCHistograms.h"
+
+DECLARE_COMPONENT(HepMCHistograms)
+
+HepMCHistograms::HepMCHistograms(const std::string& name, ISvcLocator* svcLoc) :
+    GaudiAlgorithm(name, svcLoc), m_ths(nullptr),
```
Comment (jlingema) [inline Generation/src/components/HepMCHistograms.cpp]: Slight reformat to conform with coding conventions, see above.

File: Generation/src/components/HepMCHistograms.cpp [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  
+  return StatusCode::SUCCESS;
+}
+
+StatusCode HepMCHistograms::execute() {
+  auto evt = m_hepmchandle.get();
+  
+  info() << "Processing event with " << evt->particles_size() << " particles" << endmsg;
+  
+  for (auto it = evt->particles_begin(); it != evt->particles_end(); ++it) {
```
Comment (jlingema) [inline Generation/src/components/HepMCHistograms.cpp]: A slightly better style would be (± syntax correctness)
```c++
for (auto it = evt->particles_begin(), end = evt->particles_end(); it != end; ++it) {
```

File: Generation/src/components/HepMCHistograms.cpp [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    auto particle = *it;
+    m_eta->Fill(particle->momentum().eta());
+    m_pt->Fill(particle->momentum().perp());
+  }
+
+  for(auto it = evt->vertices_begin(); it != evt->vertices_end(); ++it) {
```
Comment (jlingema) [inline Generation/src/components/HepMCHistograms.cpp]: see above

File: Generation/src/components/HepMCHistograms.cpp (hunk 10 above)
Comment (hegner) [inline Generation/src/components/HepMCHistograms.cpp]: adding a "const" would even be nicer ;-)

Comment (hegner) [inline Generation/src/components/HepMCHistograms.cpp]: note to myself: add a const-iterator :-)

File: Generation/src/components/HepMCSimpleMerge.cpp [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+#include "GaudiKernel/DeclareFactoryEntries.h"
+
+DECLARE_TOOL_FACTORY(HepMCSimpleMerge)
+
+HepMCSimpleMerge::HepMCSimpleMerge( 
+    const std::string& type,
+    const std::string& name, 
+    const IInterface* parent)
+    : GaudiTool(type, name, parent) {
```
Comment (jlingema) [inline Generation/src/components/HepMCSimpleMerge.cpp]: Slight reformat to conform with coding conventions, see above.

File: Generation/src/components/HepMCSimpleMerge.cpp [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+StatusCode HepMCSimpleMerge::initialize() {
+  StatusCode sc = GaudiTool::initialize();
+  if (!sc.isSuccess()) return sc;
+  return sc;
+}
+
+HepMC::GenEvent* HepMCSimpleMerge::merge(const std::vector<HepMC::GenEvent>& eventVector) {
+  // initial event is assumed to be signal event -- copied in full
+  HepMC::GenEvent* pileEvent = new HepMC::GenEvent(eventVector[0]);
```
Comment (jlingema) [inline Generation/src/components/HepMCSimpleMerge.cpp]: Since you delete at the end, this should either be a smart pointer or no pointer at all.

File: Generation/src/components/HepMCSimpleMerge.cpp [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+  // iterate over vertices and add them to pileEvent
+  for (auto it=eventVector.begin()+1;
```
Comment (jlingema) [inline Generation/src/components/HepMCSimpleMerge.cpp]: A slightly nicer syntax would be to use const iterator and get end only once:
```c++
for (auto it=eventVector.cbegin()+1, end = eventVector.cend(); it != end; ++it
```

File: Generation/src/components/HepMCSimpleMerge.cpp [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+HepMC::GenEvent* HepMCSimpleMerge::merge(const std::vector<HepMC::GenEvent>& eventVector) {
+  // initial event is assumed to be signal event -- copied in full
+  HepMC::GenEvent* pileEvent = new HepMC::GenEvent(eventVector[0]);
+  // iterate over vertices and add them to pileEvent
+  for (auto it=eventVector.begin()+1; 
+      it != eventVector.end();
+       ++it) {
+    HepMC::GenEvent* tmpEvent = new HepMC::GenEvent(*it);
```
Comment (jlingema) [inline Generation/src/components/HepMCSimpleMerge.cpp]: Why do you copy the event here and not use the iterator below?

File: Generation/src/components/HepMCSimpleMerge.cpp [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    for ( HepMC::GenEvent::particle_iterator p = tmpEvent->particles_begin();
+          p != tmpEvent->particles_end();
+          ++p ) {
+        // simple check if final-state particle: 
+        // has no end vertex and correct status code meaning no further decays
+        if( !(*p)->end_vertex() && (*p)->status()==1 ) {
+          HepMC::GenParticle* tmpParticle = new HepMC::GenParticle(**p);
```
Comment (jlingema) [inline Generation/src/components/HepMCSimpleMerge.cpp]: Can you add a comment about ownership here? (Same below for the GenVertex)

File: Generation/src/components/ParticleGunAlg.cpp [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+// 2014-06-23 : Benedikt Hegner (simplification for Gaudi)
+//-----------------------------------------------------------------------------
+
+/// Declaration of the Algorithm Factory
+DECLARE_ALGORITHM_FACTORY( ParticleGunAlg )
+
+/// Standard constructor, initializes variables
+ParticleGunAlg::ParticleGunAlg( const std::string& name,
+                          ISvcLocator* pSvcLocator)
+  : GaudiAlgorithm ( name , pSvcLocator ) {
```
Comment (jlingema) [inline Generation/src/components/ParticleGunAlg.cpp]: Slight reformat to conform with coding style, see above.

File: Generation/src/components/ParticleGunAlg.cpp [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+  debug() << "==> Initialise" << endmsg ;
+  return StatusCode::SUCCESS;
+}
+
+/// Main execution
+StatusCode ParticleGunAlg::execute() {
+  StatusCode sc = StatusCode::SUCCESS ;
+  Gaudi::LorentzVector theFourMomentum ;
+  Gaudi::LorentzVector origin ;
+  int thePdgId ;
```
Comment (jlingema) [inline Generation/src/components/ParticleGunAlg.cpp]: Please add a comment that this is actually retrieved from the particleGunTool. Also maybe put it directly next to the function that sets it.

File: Generation/src/components/PoissonPileUp.cpp [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+#include "GaudiKernel/DeclareFactoryEntries.h"
+#include "GaudiKernel/IRndmGenSvc.h"
+
+DECLARE_TOOL_FACTORY( PoissonPileUp )
+
+PoissonPileUp::PoissonPileUp( const std::string& type,
+                              const std::string& name,
+                              const IInterface* parent )
+  : GaudiTool ( type, name , parent ) {
```
Comment (jlingema) [inline Generation/src/components/PoissonPileUp.cpp]: Slightly reformat to conform with coding style, see above.

File: Generation/src/components/ParticleGunAlg.cpp [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+  
+  // prepare a new HepMC event
+  HepMC::GenEvent * theEvent = new HepMC::GenEvent( HepMC::Units::GEV, HepMC::Units::CM) ;
```
Comment (jlingema) [inline Generation/src/components/ParticleGunAlg.cpp]: Can you add a comment about lifetime of event, vertex and particle?

File: Generation/src/components/HepMCHistograms.cpp (hunk 10 above)
Comment (vvolkl) [inline Generation/src/components/HepMCHistograms.cpp]: Should I include that? Is there a way to get a const iterator using `auto`, i.e. without explicitly specifying `HepMC::GenEvent::particle_const_iterator`?

**Comment by vvolkl** [comment]
Many thanks for the comments!

**Comment by jlingema** [comment]
Still work in progress or ready to be merged?

File: Generation/src/components/HepMCFileReaderTool.h [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+class HepMCFileReader: public GaudiTool, virtual public IHepMCFileReaderTool {
+public:
+  HepMCFileReader( const std::string& type, const std::string& name,
+    const IInterface* parent) ;
+  virtual ~HepMCFileReader(); ///< Destructor
+  virtual StatusCode initialize();
+  virtual StatusCode finalize();
+
+  /// Wrapper for HepMC's fill_next_event() --
+  /// ownership of the event is given to the data service (in algorithm HepMCReader)
```
Comment (jlingema) [inline Generation/src/components/HepMCFileReaderTool.h]: Maybe rephrase this, as somebody might want to use the tool in another algorithm.. Something like "the user is responsible for the deletion of the event"

File: Generation/src/components/MomentumRangeParticleGun.cpp [hunk 22]
```diff
@@ -32,27 +27,23 @@ MomentumRangeParticleGun::MomentumRangeParticleGun( const std::string & type ,
   declareProperty( "MomentumMin"  , m_minMom = 100.0 * Gaudi::Units::GeV, "Minimal momentum"  ) ;
   declareProperty( "ThetaMin"     , m_minTheta = 0.1 * Gaudi::Units::rad, "Minimal theta"     ) ;
   declareProperty( "PhiMin"       , m_minPhi = 0. * Gaudi::Units::rad,    "Minimal phi"       ) ;
-    
+  
   declareProperty( "MomentumMax" , m_maxMom   = 100.0 * Gaudi::Units::GeV, "Maximal momentum" ) ;
   declareProperty( "ThetaMax"     , m_maxTheta = 0.4 * Gaudi::Units::rad,  "Maximal theta"    ) ;
   declareProperty( "PhiMax"      , m_maxPhi   = Gaudi::Units::twopi * Gaudi::Units::rad, "Maximal phi" ) ;
-
+
```
Comment (jlingema) [inline Generation/src/components/MomentumRangeParticleGun.cpp]: There are still some whitespace additions in this file.

File: Generation/src/components/MomentumRangeParticleGun.h [hunk 23]
```diff
@@ -29,39 +26,39 @@ class MomentumRangeParticleGun : public GaudiTool , virtual public IParticleGunT
   
   /// Initialize particle gun parameters
   virtual StatusCode initialize();
-
+
```
Comment (jlingema) [inline Generation/src/components/MomentumRangeParticleGun.h]: There are still some whitespace additions in this file.

**Comment by vvolkl** [comment]
Corrected to try yo avoid whitespace changes as much as possible. I think it is ready to be merged.
